client/src/js/services/ExchangeRateService.js

```diff
@@ -196,19 +213,40 @@ function ExchangeRateService($http, util, Currencies, Session) {
     // you to apply this transformation to a list of mixed currencies.
     if (currencyId === Session.enterprise.currency_id) { return 1; }
 
-    // look up the rates for currencyId via the cMap object.
-    const rates = cMap[currencyId].filter((row) => row.date <= cdate);
+    // Get the current rates
+    if (cMap[currencyId]) {
+      // look up the rates for currencyId via the cMap object
+      const rates = cMap[currencyId].filter((row) => row.date <= cdate);
 
-    // get the last rate for the given currency
-    const { rate } = rates[rates.length - 1];
+      // get the last rate for the given currency
+      const { rate } = rates[rates.length - 1];
 
-    return rate;
+      return rate;
+    }
+
+    // Warn the user if the currency does not have an exchange rate
+    Notify.danger('EXCHANGE.MUST_DEFINE_RATES_FIRST', 60000);
+    return null;
   }
 
   function del(id) {
     return $http.delete(`/exchange/${id}`)
       .then(util.unwrapHttpResponse);
   }
 
+  function getMissingExchangeRates() {
+    return service.missingRates;
```

Comment:
This list needs to be refreshed on create/update/delete.

Let's move the logic for creating the `missingRates` into a separate helper function (`computeMissingRates()` or something) that can get called when the exchange rate is updated, deleted, or created.

Reply:
Did not think a separate function was really necessary since all the necessary logic is in read() and writing a stand-alone computeMissingRates() would have duplicated a lot of read().   So I just updated getMissingExchangeRates() to call read and fixed an oversight in read().